### PR TITLE
Added asynchronous Path class

### LINKED ISSUE #296: anyio.Path like trio.Path support
could probably lift the implementation from https://github.com/python-trio/trio/blob/master/trio/_path.py but substitute anyio.to_thread.run

### PR DESCRIPTION
**Pull request by agronholm**
Fixes #296.

### CODE REVIEW

File: src/anyio/_core/_fileio.py [hunk 1]
```diff
@@ -114,3 +124,208 @@ async def open_file(file: Union[str, PathLike, int], mode: str = 'r', buffering:
...
+
+        await to_thread.run_sync(self._path.rename, target)
+
+    async def replace(self, target: Union[str, pathlib.PurePath, 'Path']) -> None:
+        if isinstance(target, Path):
+            target = target._path
+
+        await to_thread.run_sync(self._path.replace, target)
+
+    async def resolve(self) -> 'Path':
```
Comment (graingert) [inline src/anyio/_core/_fileio.py:280]: resolve has a strict kwarg

Comment (agronholm) [reply src/anyio/_core/_fileio.py:280]: Yeah, I was gonna add that, just slipped my mind. But thanks for the reminder.

**Comment by agronholm** [comment]
This PR needs a whole bunch of additional tests before it's ready.

**Comment by agronholm** [comment]
What this is still missing is some documentation of which operations require `await`.

File: src/anyio/__init__.py [hunk 2]
```diff
@@ -72,6 +73,7 @@
     'TypedAttributeProvider'
 )
 
+# isort: off
```
Comment (graingert) [inline src/anyio/__init__.py:76]: Do you still need this?

File: src/anyio/_core/_fileio.py [hunk 3]
```diff
@@ -114,3 +125,346 @@ async def open_file(file: Union[str, PathLike, int], mode: str = 'r', buffering:
...
+    __slots__ = '_path'
+
+    def __init__(self, *args: Union[str, PathLike]) -> None:
+        self._path: Final[pathlib.Path] = pathlib.Path(*args)
+
+    def __fspath__(self) -> str:
+        return self._path.__fspath__()
+
+    def __str__(self) -> str:
+        return self._path.__str__()
```
Comment (graingert) [inline src/anyio/_core/_fileio.py:202]: These dunders need to be called via the class with `type(v).__str__(v)` or just `str(v)`

File: src/anyio/_core/_fileio.py [hunk 4]
```diff
@@ -114,3 +125,346 @@ async def open_file(file: Union[str, PathLike, int], mode: str = 'r', buffering:
...
+    def with_stem(self, stem: str) -> 'Path':
+        return Path(self._path.with_name(stem + self._path.suffix))
+
+    def with_suffix(self, suffix: str) -> 'Path':
+        return Path(self._path.with_suffix(suffix))
+
+    async def write_bytes(self, data: bytes) -> int:
+        return await to_thread.run_sync(self._path.write_bytes, data)
+
+    async def write_text(self, data: str, encoding: Optional[str] = None,
```
Comment (graingert) [inline src/anyio/_core/_fileio.py:464]: Why not to_thread the write_text directly?

File: src/anyio/_core/_fileio.py (hunk 3 above)
Comment (smurfix) [reply src/anyio/_core/_fileio.py:202]: I don't understand this comment. `str(x)` or `x.__str__()` or `type(x).__str__(x)` are all equivalent.

Comment (graingert) [reply src/anyio/_core/_fileio.py:202]: @smurfix they're different:

```python
class Foo:
    def __init__(self) -> None:
        self.__str__ = lambda: "x.__str__()"

    def __str__(self) -> str:
        return "type(x).__str__(x)"

print(f"{str(Foo())=}")  # str(Foo())='type(x).__str__(x)'
print(f"{Foo().__str__()=}")   # Foo().__str__()='x.__str__()'
```

File: src/anyio/_core/_fileio.py (hunk 4 above)
Comment (agronholm) [reply src/anyio/_core/_fileio.py:464]: Because on py < 3.10, the `newline` parameter doesn't exist.

File: src/anyio/_core/_fileio.py (hunk 3 above)
Comment (smurfix) [reply src/anyio/_core/_fileio.py:202]: Yes I know that `str()` takes a shortcut and doesn't do a lookup on the instance. But which real-world class assigns to `self.__str__`? `pathlib.Path`certainly doesn't.
```
>>> import pathlib as p
>>> x=p.Path("a/b/c")
>>> x
PosixPath('a/b/c')
>>> str(x)
'a/b/c'
>>> x.__str__()
'a/b/c'
>>> 
>>> type(x).__str__(x)
'a/b/c'
>>>
```

Comment (agronholm) [reply src/anyio/_core/_fileio.py:202]: I'm also not understanding why this needs to be changed. Is there a potential problem with this way of doing things?

File: src/anyio/__init__.py (hunk 2 above)
Comment (agronholm) [reply src/anyio/__init__.py]: Apparently not – I've removed it now.

File: src/anyio/_core/_fileio.py [hunk 5]
```diff
@@ -114,3 +125,346 @@ async def open_file(file: Union[str, PathLike, int], mode: str = 'r', buffering:
...
+    * :meth:`~pathlib.Path.write_text`
+
+    Additionally, the following methods return an async iterator yielding :class:`~.Path` objects:
+
+    * :meth:`~pathlib.Path.glob`
+    * :meth:`~pathlib.Path.iterdir`
+    * :meth:`~pathlib.Path.rglob`
+    """
+
+    __slots__ = '_path'
```
Comment (graingert) [inline src/anyio/_core/_fileio.py:193]:
```suggestion
    __slots__ = '_path', '__weakref__'
```

**Review by graingert — CHANGES_REQUESTED** [review]

File: src/anyio/_core/_fileio.py (hunk 4 above)
Comment (graingert) [reply src/anyio/_core/_fileio.py:464]: Would it be better to do this in one call to to_thread?

File: src/anyio/_core/_fileio.py (hunk 3 above)
Comment (graingert) [reply src/anyio/_core/_fileio.py:202]: Ah I forgot pathlib.Path was frozen. It's usually a problem but not in this specific case

File: src/anyio/_core/_fileio.py (hunk 5 above)
Comment (agronholm) [reply src/anyio/_core/_fileio.py:193]: The `pathlib.Path` instance is not weak referable. That is why I didn't include this. I wonder if the lack of a `__weakref__` slot is intentional there.

Comment (graingert) [reply src/anyio/_core/_fileio.py:193]: Oh that's very surprising

Comment (graingert) [reply src/anyio/_core/_fileio.py:193]: Sounds like a cPython bug that anyio should copy. Might be worth a test

Comment (agronholm) [reply src/anyio/_core/_fileio.py:193]: Interestingly, on PyPy3 they *are* weakrefable. I don't think there is any harm done adding the weakref slot, so I'll do that.

Comment (graingert) [reply src/anyio/_core/_fileio.py:193]: Pypy makes everything weakrefable

Comment (agronholm) [reply src/anyio/_core/_fileio.py:193]: I'm not sure what to do about this, but I don't think we should test that it's *not* weakrefable.

File: src/anyio/_core/_fileio.py (hunk 4 above)
Comment (agronholm) [reply src/anyio/_core/_fileio.py:464]: Done

**Review by graingert — APPROVED** [review]
Nice

**Comment by agronholm** [comment]
Thanks for the review! As a side note, we really need to get rid of the flakiness of the TLS tests on Windows.

**Pull request merged**